zuul-core/src/main/java/com/netflix/zuul/netty/filter/BaseZuulFilterRunner.java

```diff
@@ -171,10 +171,13 @@ protected final O filter(final ZuulFilter<I, O> filter, final I inMesg) {
                 return (outMesg != null) ? outMesg : filter.getDefaultOutput(inMesg);
             }
 
-            //async filter
+            // async filter
+            filter.incrementConcurrency();
+            final FilterChainResumer resumer = new FilterChainResumer(inMesg, filter, snapshot, startTime);
             filter.applyAsync(inMesg)
-                    .observeOn(Schedulers.from(getChannelHandlerContext(inMesg).executor()))
-                    .subscribe(new FilterChainResumer(inMesg, filter, snapshot, startTime));
+                .observeOn(Schedulers.from(getChannelHandlerContext(inMesg).executor()))
+                .doOnUnsubscribe(resumer::decrementConcurrency)
```

Comment:
To clarify, this unsubscribe will be called after this individual async filter completes/errors, correct?

Reply:
"If" it is called. I don't think we actually unsubscribe the observer in normal case. This is paranoid safety measure so that no concurrency count may leak. Note that it will be safe for it (decrementConcurrency) to be called multiple times as FilterChainResumer internally keeps state of decrement operation and does it only once on first decrement.

Reply:
I was more concerned about this counter being incremented/decremented per filter instead of the entire filter chain.

Reply:
Ah, yes, the FilterChainResumer is per AsyncFilter. So, yes, the counter will get incremented/decremented per individual filter execution.